### ISSUE TITLE
Improve gawk support

### ISSUE THREAD

**Issue by jmdevin** [vim/vim#15708]
# Preface
I'm willing (and I think, able) to implement this solution myself when I have the time, so this is more of a request for review and ideas from the maintainer (@dkearns) and anyone else who could offer some insight before I spend more time on this.

# Issue
Right now, the awk ftplugin ([link](https://github.com/vim/vim/blob/master/runtime/ftplugin/awk.vim)) allows the user to set `g:awk_is_gawk` to add some additional support for the gnu version of awk. It also conveniently defines `'formatprg'` to use gawk's built-in pretty-printer.

There are two main issues using gawk as `'formatprg'`:
## 1. Tabs
gawk's `--pretty-print` or `-o` option ~~seems to output~~ outputs exclusively tab-indented code ~~(I've emailed the gawk help mailing list to check, I'll update this if I'm wrong)~~ (update from maintainer: yes, tabs only) . This behavior fails to respect the user's `'expandtab'` setting.

## 2. Errors
gawk's pretty-printer will emit only an error message when it encounters a syntax error. This can lead to the user's entire file being replaced by a two-line error message after a simple `gggqG`.

# Suggested Solution
I *believe* that using `'formatexpr'` to call gawk instead of `'formatprg'` would let us address both of these issues. `'formatexpr'` should be able to replace the tabs output by gawk with an appropriate number of spaces according to `'expandtab'` and `'tabstop'`. It should also be able to do an `echoerr` to warn the user and simply replace the faulty text when gawk responds with an error message, rather than simply replacing all the user's text with an error.

## Alternatives Considered
Originally, I had no solution for the error problem, and I was simply going to make a pull request to add `noexpandtab` when the user defined `g:awk_is_gawk`. This would be similar to the filetype plugin for go using `g:go_recommended_style` (see here: [link](https://github.com/vim/vim/blob/master/runtime/ftplugin/go.vim)).

However, the go plugin can justify that, as golang's opinionated formatting has been enforced since it's inception. Adding `noexpandtab` for awk based on the user defining that variable would be unusual. Its effect also isn't visually obvious, meaning that a user who's unaware of it might screw up a lot of their files before noticing.

# Additional context
The second edition of *The AWK Programming Language* came out just last year, so if you have some spare change buy a copy and make sure good programming books can still earn money: [link](https://awk.dev/). I don't get anything from it, I just like good books.

**Comment by chrisbra** [comment]
Thanks for offering help. But I am not sure if `'formatexpr'` can fix this issue easily. Using formatexpr itsself can be a bit tricky as well. But perhaps you can test the return status of gawk and if there was an error just roll back the change. That should be doable I think.

**Comment by zzzyxwvut** [comment]
> gawk's --pretty-print or -o option seems to output exclusively tab-indented code (I've emailed the gawk help mailing list to check, I'll update this if I'm wrong). This behavior fails to respect the user's 'expandtab' setting.

There is [`expand(1)`](https://www.gnu.org/software/coreutils/manual/coreutils.html#expand-invocation) when runs of spaces are preferred.

E.g. `~/.vim/after/ftplugin/awk.vim`:

```vim
if exists("g:awk_is_gawk")
    let &l:fp .= ' | expand -t' . &l:sw
endif
```

**Comment by jmdevin** [comment]
> Thanks for offering help. But I am not sure if `'formatexpr'` can fix this issue easily. Using formatexpr itsself can be a bit tricky as well. But perhaps you can test the return status of gawk and if there was an error just roll back the change. That should be doable I think.

Yes, I didn't realize how few of vim's ftplugins were using `'formatexpr'` - seems like only three. Luckily, one of them is [https://github.com/chrisbra/vim-xml-runtime](https://github.com/chrisbra/vim-xml-runtime), which served as a decent guide to shamelessly copy from (thanks for it).

I have got a proof-of-concept just about working; it only needs a few changes to handle whitespace better when the user formats a range instead of the whole file. I'll post it here soon for comments.

> There is [`expand(1)`](https://www.gnu.org/software/coreutils/manual/coreutils.html#expand-invocation) when runs of spaces are preferred.

I had no idea that utility existed. I probably won't use it in the ftplugin just to avoid the dependency, but you actually just solved a separate problem I was having, so thank you.

**Comment by zzzyxwvut** [comment]
Another one (`autoload/javaformat.vim`) is written and will be  
included for the support of [Markdown documentation comments](https://openjdk.org/jeps/467#Syntactical-details)  
in Java source files.  Although, no *automatic* line rewriting  
will be offered out of the box.

**Comment by jmdevin** [comment]
> This is really a general problem when using `'formatprg'`. The Modula-3 and Racket formatters exhibit the same behaviour. It's not currently possible to use any of them without understanding the details of the external tool. Perhaps it would be best to eventually generalise this to a shared plugin?

Yes, I had the same thought while starting to write this. I can't put any work into it right now, but I might be able to in the future if someone more knowledgeable could assist in speccing it out.

Just as a note, the maintainer of gawk is quite responsive. He's making some minor improvements to gawk's pretty-printer, and already has one patch done. They won't obviate the need for the improvements I mentioned above, but it's definitely appreciated.

> Thanks for working on this. I cobbled something similar together for the M3 pretty printer a few years ago but, unsurprisingly, I can't seem to find it right now.

Don't thank me before you see what kind of code I write ;)

**Comment by Konfekt** [comment]
> gawk's pretty-printer will emit only an error message when it encounters a syntax error. This can lead to the user's entire file being replaced by a two-line error message after a simple gggqG.

This is a common ftplugin idiom

```vim
  if exists('##ShellFilterPost')
    autocmd ShellFilterPost <buffer> if v:shell_error | execute 'echom "shell filter returned error " . v:shell_error . ", undoing changes"' | undo | endif
  endif
```

to keep the text despite errors
